### DISCUSSION TITLE
Making View::Pick Sync?

### DISCUSSION THREAD

**Discussion by bjovanovic84** [google/filament#5799] [General]
We have an interesting problem at hand: picking functionality on `View` in async (using the callback) and on most regular platforms (macos, ios, android) we were able to make it sync by performing rendering cycles until it becomes available, something like:
```
auto result_ready = std::atomic<bool>(false);
auto result = View::PickingQueryResult();
auto on_picked = [&](const auto &picked) {
  result = picked;
  result_ready.store(true);
};
view->pick(x, y, on_picked);
if (renderer->beginFrame(swap_chain, now)) {
  renderer->render(view_);
  renderer->endFrame();
}
engine->flushAndWait();

// result is available here - we have a loop that waits for it too, just to be safe

```

However, the approach above does not work for webgl, the very similar example (with `engine->update()` instead of `engine->flushAndWait()`) does not work, even with 100k rendering iterations. The callback does eventually get called from a separate rendering call from the js side, furthermore it seems like it is an implicit requirement.

Question: is there a way to make picking be sync for webgl? Are there some internal bells and whistles that we could tweak on the filament side to make this happen (fyi, we already have a fork, not a problem to make the change even if it's buried deep)?

Note: we could make use of async picking, but it severely limits the functionality for our intended usage, we would rather avoid it if it's at all possible.

**Comment by romainguy** [comment]
Why do you need pick to be synchronous? The problem with making is synchronous is that you have to block your thread and wait on the GPU to be done.

**Reply by bjovanovic84** [reply]
We have multiple extensions active in the scene that respond to viewport clicks and do different things depending on the picking outcome, i.e. if a pick is successful then add something to the scene, if it's not pass the click event to the next tool to do something different etc. Having pick async is making it very hard, waiting for the gpu is fine, `flushAndWait` is fairly quick and the block is not really noticeable to the users

**Reply by bjovanovic84** [reply]
one additional complication is using layers with picking, w/o affecting what the users see - we want all the layers to be visible to the users, but perform picking using only a subset

**Reply by cpheinrich** [reply]
> Why do you need pick to be synchronous? The problem with making is synchronous is that you have to block your thread and wait on the GPU to be done.

Side stepping the issue of our motivation for making it synchronous --- is this something you think we could do by modifying the current implementation of picking? If so, do you have any suggestions for how to go about this? As Bojan noted, waiting for the GPU is much less of an issue than designing around having picking async. Many renderers (SceneKit and ThreeJS to name a few) provide synchronous picking, so it does seem like a reasonable request

**Reply by pixelflinger** [reply]
No there is no way to make it synchronous. Picking can only be started from within `Renderer.render()`, at the point of `pick` is called, we're not in a state where we can perform the picking pass.
So at the very least you'd have to wait for Renderer.render() do be called.

I think your solution with `flushAndWait` doesn't work on WebGL, maybe because you need a call to Engine::execute() before the `flushAndWait`. I think on WebGL you need to manage the calls to `execute()` yourself.

**Reply by bjovanovic84** [reply]
we’re using execute (I made a typo and said update instead of it), but it does not work. the callback gets triggered in the next render pass as soon as we release the thread and re-enter the rendering cycle, almost feels like that there is no way to force the command queue to drain for webgl. somebody with knowledge about the webgl backend implementation might be able to shed some light
